### PR TITLE
adding dcache,icache,rocket

### PR DESCRIPTION
**Pull request by depappas-sifive**
Adding object models for dcache, icache, and rocket

feature request

API addition (no impact on existing code)

implementation

**Release Notes**

### CODE REVIEW

File: src/main/scala/tile/RocketTile.scala [hunk 1]
```diff
@@ -80,6 +81,111 @@ class RocketTile(
...
+          nRegions = coreParams.nPMPs,
+          granularity = coreParams.pmpGranularity
+        ))
+      }
+      else { None }
+
+      val mulDiv = coreParams.mulDiv.map{ md => OMMulDiv.makeOMI(md, xLen)}
+
+      val baseInstructionSet = xLen match {
+        case 32 => if (XLen == 32) RV32E else RV32I // TODO coreParams.useRVE
```
Comment (richardxia) [inline src/main/scala/tile/RocketTile.scala]: Why is this here? Rocket doesn't support RVE.

File: src/main/scala/tile/RocketTile.scala [hunk 2]
```diff
@@ -80,6 +81,111 @@ class RocketTile(
...
+        xLen = xLen,
+        baseSpecification = baseSpec(baseInstructionSet, baseISAVersion),
+        base = baseInstructionSet,
+        m = coreParams.mulDiv.map { x => isaExtSpec(M, "2.0") },
+        a = coreParams.useAtomics.option(isaExtSpec(A, "2.0")),
+        f = coreParams.fpu.map { x => isaExtSpec(F, "2.0") },
+        d = d,
+        c = coreParams.useCompressed.option(isaExtSpec(C," 2.0")),
+        u = coreParams.useUser.option(isaExtSpec(U,"1.10")),
+        s = None,
```
Comment (richardxia) [inline src/main/scala/tile/RocketTile.scala]: This isn't correct. A Rocket core may support S mode depending on the parameters passed in.

File: src/main/scala/tile/RocketTile.scala [hunk 3]
```diff
@@ -80,6 +81,111 @@ class RocketTile(
...
+        addressTranslationModes = Nil
```
Comment (richardxia) [inline src/main/scala/tile/RocketTile.scala]: Similarly, it will support certain address translation modes if S mode is enabled.

File: src/main/scala/tile/RocketTile.scala [hunk 4]
```diff
@@ -80,6 +81,111 @@ class RocketTile(
...
+      )
+
+      val btb = rocketParams.btb.map(BTB.makeOMI(_))
+
+      val omICache = frontend.icache.device.getOMComponents(resourceBindingsMap).headOption.getOrElse(None).asInstanceOf[OMICache]
```
Comment (richardxia) [inline src/main/scala/tile/RocketTile.scala]: I'm not sure I understand what's going on here. What does it mean to `.getOrElse(None)`? That ends up returning either an `A` or a `None` rather than a `Some[A]`, which is a very odd type to return.

File: src/main/scala/tile/RocketTile.scala [hunk 5]
```diff
@@ -80,6 +81,111 @@ class RocketTile(
...
+
+      def dcache(p: DCacheParams): OMDCache = {
+        OMDCache(
+          memoryRegions = Nil,
```
Comment (richardxia) [inline src/main/scala/tile/RocketTile.scala]: If the DCache is configured as a DTIM, then you should populate the memory regions here.

File: src/main/scala/tile/RocketTile.scala [hunk 6]
```diff
@@ -80,6 +81,111 @@ class RocketTile(
...
+      Seq(OMRocketCore(
+        isa = omIsa,
+        mulDiv = mulDiv,
+        fpu = fpu,
+        performanceMonitor = perfMon,
+        pmp = pmp,
+        documentationName = "TODO",
+        hartIds = Seq(hartId),
+        hasVectoredInterrupts = true,
+        interruptLatency = 6,
```
Comment (richardxia) [inline src/main/scala/tile/RocketTile.scala]: Where did this number come from and is it even correct?

File: src/main/scala/tile/RocketTile.scala [hunk 7]
```diff
@@ -80,6 +81,111 @@ class RocketTile(
...
+        nLocalInterrupts = coreParams.nLocalInterrupts,
+        nBreakpoints = coreParams.nBreakpoints,
+        branchPredictor = btb,
+        dcache = omDCache,
+        icache = Some(omICache)
```
Comment (richardxia) [inline src/main/scala/tile/RocketTile.scala]: Why is the ICache a Option but the DCache is not?

File: src/main/scala/subsystem/RocketSubsystem.scala [hunk 8]
```diff
@@ -53,6 +54,15 @@ trait HasRocketTiles extends HasTiles
 
     rocket
   }
+
+  abstract override def getSubsystemOMComponents(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] = {
+    val rockets = getOMRocketCores(resourceBindingsMap)
+    val plics = plicOpt.map(_.device.getOMComponents(resourceBindingsMap))
+    val clints = clintOpt.map(_.device.getOMComponents(resourceBindingsMap)) // super.getSubsystemOMComponents(resourceBindingsMap)
```
[the remark this answers is not in the document]
Comment (depappas-sifive) [reply src/main/scala/subsystem/RocketSubsystem.scala]: No

File: src/main/scala/diplomaticobjectmodel/model/OMCaches.scala [hunk 9]
```diff
@@ -2,6 +2,10 @@
 
 package freechips.rocketchip.diplomaticobjectmodel.model
 
+import freechips.rocketchip.diplomacy.ResourceBindings
+import freechips.rocketchip.diplomaticobjectmodel.DiplomaticObjectModelAddressing
+import freechips.rocketchip.rocket.{DCacheParams, ICacheParams}
+
```
[the remark this answers is not in the document]
Comment (depappas-sifive) [reply src/main/scala/diplomaticobjectmodel/model/OMCaches.scala]: removed

File: src/main/scala/rocket/ICache.scala [hunk 10]
```diff
@@ -54,7 +56,34 @@ class ICache(val icacheParams: ICacheParams, val hartId: Int)(implicit p: Parame
...
   val size = icacheParams.nSets * icacheParams.nWays * icacheParams.blockBytes
-  val device = new SimpleDevice("itim", Seq("sifive,itim0"))
+  val device = new SimpleDevice("itim", Seq("sifive,itim0")) {
+    override def getOMComponents(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] = {
+      require(resourceBindingsMap.map.contains(this))
+      val resourceBindings = resourceBindingsMap.map.get(this)
+      resourceBindings.map {
+        rb => icache(icacheParams, rb)
+      }
+    }.getOrElse(throw new NoSuchElementException)
```
[the remark this answers is not in the document]
Comment (depappas-sifive) [reply src/main/scala/rocket/ICache.scala]: moved inside of the braces

File: src/main/scala/rocket/ICache.scala [hunk 11]
```diff
@@ -54,7 +56,34 @@ class ICache(val icacheParams: ICacheParams, val hartId: Int)(implicit p: Parame
...
+  }
+
+  def icache(p: ICacheParams, resourceBindings: ResourceBindings): Seq[OMComponent] = {
```
[the remark this answers is not in the document]
Comment (depappas-sifive) [reply src/main/scala/rocket/ICache.scala]: renamed

File: src/main/scala/rocket/ICache.scala [hunk 12]
```diff
@@ -54,7 +56,34 @@ class ICache(val icacheParams: ICacheParams, val hartId: Int)(implicit p: Parame
...
+    val regions = DiplomaticObjectModelAddressing.getOMMemoryRegions("ICache", resourceBindings)
+    Seq[OMComponent](
+      OMICache(
+        memoryRegions = DiplomaticObjectModelAddressing.getOMMemoryRegions("ICache", resourceBindings),
```
[the remark this answers is not in the document]
Comment (depappas-sifive) [reply src/main/scala/rocket/ICache.scala]: removed

File: src/main/scala/rocket/ICache.scala [hunk 13]
```diff
@@ -54,7 +56,34 @@ class ICache(val icacheParams: ICacheParams, val hartId: Int)(implicit p: Parame
...
+        interrupts = Nil,
+        nSets = p.nSets,
+        nWays = p.nWays,
+        blockSizeBytes = p.blockBytes,
+        dataMemorySizeBytes = p.nSets * p.nWays * p.blockBytes,
+        dataECC = p.dataECC.map{case code => OMECC.getCode(code)},
+        tagECC = p.tagECC.map{case code => OMECC.getCode(code)},
```
[the remark this answers is not in the document]
Comment (depappas-sifive) [reply src/main/scala/rocket/ICache.scala]: fixed

File: src/main/scala/tile/RocketTile.scala (hunk 7 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: case class OMRocketCore(
  dcache: Option[OMDCache],
  icache: Option[OMICache]

File: src/main/scala/tile/RocketTile.scala (hunk 1 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: comment removed

File: src/main/scala/tile/RocketTile.scala [hunk 14]
```diff
@@ -80,6 +81,111 @@ class RocketTile(
...
+      val icache = getOMICacheFromBindings(resourceBindingsMap)
+      cores ++: icache
+    }
+    def getOMICacheFromBindings(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] = {
+      frontend.icache.device.getOMComponents(resourceBindingsMap)
+    }
+    def getOMRocketCores(resourceBindingsMap: ResourceBindingsMap): Seq[OMRocketCore] = {
+      val coreParams = rocketParams.core
+
+      val fpu= coreParams.fpu.map{f => OMFPU(fLen = f.fLen)}
```
[the remark this answers is not in the document]
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: added.

File: src/main/scala/tile/RocketTile.scala (hunk 6 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: @hcook says we should remove the interruptLatency number from the docs.

File: src/main/scala/tile/RocketTile.scala (hunk 4 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: fixed

File: src/main/scala/tile/RocketTile.scala (hunk 3 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: I was told by @hcook and @jsmithsf to use useVM to condition the S mode.

File: src/main/scala/rocket/ICache.scala [hunk 15]
```diff
@@ -54,7 +56,33 @@ class ICache(val icacheParams: ICacheParams, val hartId: Int)(implicit p: Parame
...
+    override def getOMComponents(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] = {
+      require(resourceBindingsMap.map.contains(this))
+      val resourceBindings = resourceBindingsMap.map.get(this)
+      resourceBindings.map {
+        rb => makeOMI(icacheParams, rb)
+      }.getOrElse(throw new NoSuchElementException)
+    }
+  }
+
+  def makeOMI(p: ICacheParams, resourceBindings: ResourceBindings): Seq[OMComponent] = {
```
Comment (hcook) [inline src/main/scala/rocket/ICache.scala]: What is the rationale behind making this a separate function when it is only called once?

Comment (hcook) [reply src/main/scala/rocket/ICache.scala]: Maybe this should be handled like the BTB, where it is a method on a helper object in the `diplomaticobjectmodel.model` package? I like that pattern

File: src/main/scala/diplomaticobjectmodel/model/OMRocketCore.scala [hunk 16]
```diff
@@ -26,3 +28,13 @@ case class OMRocketCore(
   icache: Option[OMICache],
   _types: Seq[String] = Seq("OMRocketCore", "OMCore", "OMComponent", "OMCompoundType")
 ) extends OMCore
+
+object BTB {
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/model/OMRocketCore.scala]: OMBTB?

File: src/main/scala/subsystem/RocketSubsystem.scala [hunk 17]
```diff
@@ -53,6 +54,15 @@ trait HasRocketTiles extends HasTiles
...
     rocket
   }
+
+  abstract override def getSubsystemOMComponents(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] = {
+    val rockets = getOMRocketCores(resourceBindingsMap)
+    val plics = plicOpt.map(_.device.getOMComponents(resourceBindingsMap))
+    val clints = clintOpt.map(_.device.getOMComponents(resourceBindingsMap))
+    rockets ++ plics.getOrElse(Nil) ++ clints.getOrElse(Nil)
+  }
+  def getOMRocketCores(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] =
```
Comment (hcook) [inline src/main/scala/subsystem/RocketSubsystem.scala]: Same could use-a-helper-method/only-called-once questions as above

File: src/main/scala/tile/RocketTile.scala (hunk 3 above)
Comment (hcook) [reply src/main/scala/tile/RocketTile.scala]: Translation modes are an addition consideration, should not just be Nil

File: src/main/scala/tile/RocketTile.scala (hunk 6 above)
Comment (hcook) [reply src/main/scala/tile/RocketTile.scala]: Yeah I have no idea what this number means

File: src/main/scala/tile/RocketTile.scala [hunk 18]
```diff
@@ -80,6 +81,117 @@ class RocketTile(
...
+      val d = coreParams.fpu.filter(_.fLen > 32).map(x => isaExtSpec(D, "2.0"))
+
+      val omIsa = OMISA(
+        xLen = xLen,
+        baseSpecification = baseSpec(baseInstructionSet, baseISAVersion),
+        base = baseInstructionSet,
+        m = coreParams.mulDiv.map { x => isaExtSpec(M, "2.0") },
+        a = coreParams.useAtomics.option(isaExtSpec(A, "2.0")),
+        f = coreParams.fpu.map { x => isaExtSpec(F, "2.0") },
+        d = d,
```
Comment (hcook) [inline src/main/scala/tile/RocketTile.scala]: Why is only `d` its own val?

File: src/main/scala/tile/RocketTile.scala (hunk 3 above)
Comment (hcook) [reply src/main/scala/tile/RocketTile.scala]: @aswaterman is there an straightforward way to calculate the mode from `coreParams`?

File: src/main/scala/tile/RocketTile.scala [hunk 19]
```diff
@@ -80,6 +81,117 @@ class RocketTile(
...
+        addressTranslationModes = Nil
+      )
+
+      val btb = rocketParams.btb.map(BTB.makeOMI(_))
+
+      val omICache = rocketParams.icache.map(i => frontend.icache.device.getOMComponents(resourceBindingsMap) match {
+        case Seq() => throw new IllegalArgumentException
+        case Seq(h) => h.asInstanceOf[OMICache]
+        case _ => throw new IllegalArgumentException
+      })
```
Comment (hcook) [inline src/main/scala/tile/RocketTile.scala]: I'm confused about the relationship between `omICache`, `getOMICacheFromBindings` and `frontend.icache.device.getOMComponents`

File: src/main/scala/tile/RocketTile.scala [hunk 20]
```diff
@@ -80,6 +81,117 @@ class RocketTile(
       val Description(name, mapping) = super.describe(resources)
       Description(name, mapping ++ cpuProperties ++ nextLevelCacheProperty ++ tileProperties ++ dtimProperty ++ itimProperty)
     }
+
+    override def getOMComponents(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] = {
+      val cores = getOMRocketCores(resourceBindingsMap)
+      val icache = getOMICacheFromBindings(resourceBindingsMap)
```
Comment (hcook) [inline src/main/scala/tile/RocketTile.scala]: Why are icache and dcache handled differently here?

File: src/main/scala/tile/RocketTile.scala [hunk 21]
```diff
@@ -80,6 +81,117 @@ class RocketTile(
...
+      val btb = rocketParams.btb.map(BTB.makeOMI(_))
+
+      val omICache = rocketParams.icache.map(i => frontend.icache.device.getOMComponents(resourceBindingsMap) match {
+        case Seq() => throw new IllegalArgumentException
+        case Seq(h) => h.asInstanceOf[OMICache]
+        case _ => throw new IllegalArgumentException
+      })
+
+//      //If DCache is configured as a DTIM, then you should populate the memory regions
+      def dcache(p: DCacheParams): OMDCache = {
```
Comment (hcook) [inline src/main/scala/tile/RocketTile.scala]: Similarly, seems like a good candidate for a method on an `object OMDCache` factory method, maybe
`def make OMI(p: DCacheParams, device: OMComponent)` or something like that?

File: src/main/scala/rocket/ICache.scala (hunk 15 above)
Comment (depappas-sifive) [reply src/main/scala/rocket/ICache.scala]: Made the method on a helper object in the diplomaticobjectmodel.model package

File: src/main/scala/tile/RocketTile.scala (hunk 3 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: What is the equation for addressTranslationModes?

File: src/main/scala/tile/RocketTile.scala (hunk 5 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: I spent about an hour looking for the DTIM device but could not find it.

File: src/main/scala/diplomaticobjectmodel/model/OMRocketCore.scala (hunk 16 above)
Comment (depappas-sifive) [reply src/main/scala/diplomaticobjectmodel/model/OMRocketCore.scala]: changed

File: src/main/scala/tile/RocketTile.scala (hunk 18 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: moved the val d into line 143

File: src/main/scala/tile/RocketTile.scala (hunk 19 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: I copied the pattern from line 75

      val omICache = rocketParams.icache.map(i => frontend.icache.device.getOMComponents(resourceBindingsMap) match {
        case Seq() => throw new IllegalArgumentException
        case Seq(h) => h.asInstanceOf[OMICache]
        case _ => throw new IllegalArgumentException
      })

line 75
  val itimProperty = tileParams.icache.flatMap(_.itimAddr.map(i => Map(
    "sifive,itim" -> frontend.icache.device.asProperty))).getOrElse(Nil)

File: src/main/scala/tile/RocketTile.scala (hunk 3 above)
Comment (richardxia) [reply src/main/scala/tile/RocketTile.scala]: I believe that Rocket does not have an implementation of Sv48, so the rule is:

if Xlen == 32 then Sv32
elif Xlen == 64 then Sv39

Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: fixed

File: src/main/scala/tile/RocketTile.scala (hunk 6 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: Should interruptLatency be removed from the schema?

File: src/main/scala/tile/RocketTile.scala (hunk 21 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: fixed

File: src/main/scala/tile/RocketTile.scala (hunk 20 above)
Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: fixed

File: src/main/scala/subsystem/RocketSubsystem.scala (hunk 17 above)
Comment (depappas-sifive) [reply src/main/scala/subsystem/RocketSubsystem.scala]: This is a collection point for the lower level OM*'s. 
I feel that a helper function would make the code less clear in this case.

File: src/main/scala/diplomaticobjectmodel/model/OMISA.scala [hunk 22]
```diff
@@ -36,3 +39,50 @@ case class OMISA(
...
+object OMISA {
+  // TODO the package object util has this code but is not seen in this scope
+  implicit class BooleanToAugmentedBoolean(val x: Boolean) extends AnyVal {
+    // this one's snagged from scalaz
+    def option[T](z: => T): Option[T] = if (x) Some(z) else None
+  }
+
+  def rocketISA(coreParams: RocketCoreParams, xLen: Int): OMISA = {
+    val baseInstructionSet = xLen match {
+      case 32 => if (XLen == 32) RV32E else RV32I
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/model/OMISA.scala]: Rocket *does not* currently implement RV32E, and therefore this should never return RV32E.

File: src/main/scala/tile/RocketTile.scala (hunk 6 above)
Comment (richardxia) [reply src/main/scala/tile/RocketTile.scala]: No, this is important information to capture. The value should be 4 for rocket-chip.

Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: @hcook says that the number varies depending on the interrupt set. Lets go with 4 and then in rev2 lets revisit the way that the interrupt latency's are tracked in the OM.

Comment (richardxia) [reply src/main/scala/tile/RocketTile.scala]: I just talked to @terpstra and @dbarbi1 and showed them this thread, and we think there's some confusion over what this value is supposed to be. This value is supposed to measure the latency between when an interrupt arrives at a core to when the core reacts to it, not the latency from the interrupt slaves to the core/interrupt controller.

Comment (depappas-sifive) [reply src/main/scala/tile/RocketTile.scala]: How about if we change the name of this variable to interruptCoreResponseLatency?
Is the latency 4 for rocket?
What is the latency for Caboose?

File: src/main/scala/diplomaticobjectmodel/model/OMISA.scala (hunk 22 above)
Comment (depappas-sifive) [reply src/main/scala/diplomaticobjectmodel/model/OMISA.scala]: fixed

File: src/main/scala/diplomaticobjectmodel/model/OMCaches.scala (hunk 9 above)
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/model/OMCaches.scala]: These imports aren't used anymore.

File: src/main/scala/diplomaticobjectmodel/model/OMISA.scala [hunk 23]
```diff
@@ -2,6 +2,10 @@
 
 package freechips.rocketchip.diplomaticobjectmodel.model
 
+import freechips.rocketchip.rocket.RocketCoreParams
+import freechips.rocketchip.tile.XLen
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/model/OMISA.scala]: This import isn't used.

Comment (depappas-sifive) [reply src/main/scala/diplomaticobjectmodel/model/OMISA.scala]: fixed

File: src/main/scala/diplomaticobjectmodel/model/OMCaches.scala (hunk 9 above)
Comment (depappas-sifive) [reply src/main/scala/diplomaticobjectmodel/model/OMCaches.scala]: fixed
